diff --git setup.py setup.py
index 8099299..3b4baff 100644
--- setup.py
+++ setup.py
@@ -36,6 +36,7 @@
 
 __revision__ = "$Id$"
 
+import setuptools
 from distutils import core
 from distutils.ccompiler import new_compiler
 from distutils.core import Extension, Command
@@ -390,6 +391,7 @@ kw = {'name':"pycrypto",
       'author':"Dwayne C. Litzenberger",
       'author_email':"dlitz@dlitz.net",
       'url':"http://www.pycrypto.org/",
+      'zip_safe':False,
 
       'cmdclass' : {'build_configure': PCTBuildConfigure, 'build_ext': PCTBuildExt, 'build_py': PCTBuildPy, 'test': TestCommand },
       'packages' : ["Crypto", "Crypto.Hash", "Crypto.Cipher", "Crypto.Util",
